File: app/src/main/java/tech/pauly/findapet/discover/AnimalListItemViewModel.java
package tech.pauly.findapet.discover;

import android.databinding.ObservableBoolean;
import android.databinding.ObservableField;

import java.util.List;

import javax.inject.Inject;

import tech.pauly.findapet.data.models.Animal;
import tech.pauly.findapet.data.models.Contact;
import tech.pauly.findapet.data.models.Media;
import tech.pauly.findapet.data.models.Photo;
import tech.pauly.findapet.data.models.PhotoSize;
import tech.pauly.findapet.shared.BaseViewModel;
import tech.pauly.findapet.shared.LocationHelper;
import tech.pauly.findapet.shared.ResourceProvider;
import tech.pauly.findapet.shared.datastore.AnimalDetailsUseCase;
import tech.pauly.findapet.shared.datastore.TransientDataStore;
import tech.pauly.findapet.shared.events.ActivityEvent;
import tech.pauly.findapet.shared.events.ViewEventBus;

public class AnimalListItemViewModel extends BaseViewModel {

    public ObservableField<String> name = new ObservableField<>("");
    public ObservableField<String> imageUrl = new ObservableField<>("");
    public ObservableField<String> age = new ObservableField<>("");
    public ObservableField<String> breeds = new ObservableField<>("");
    public ObservableField<String> distance = new ObservableField<>("?");
    public ObservableBoolean distanceVisibility = new ObservableBoolean(false);

    private Animal animal;
    private ViewEventBus eventBus;
    private TransientDataStore dataStore;
    private LocationHelper locationHelper;

    protected AnimalListItemViewModel(Animal animal,
                                      ViewEventBus eventBus,
                                      TransientDataStore dataStore,
                                      ResourceProvider resourceProvider,
                                      LocationHelper locationHelper) {
        this.animal = animal;
        this.eventBus = eventBus;
        this.dataStore = dataStore;
        this.locationHelper = locationHelper;
        populateFields(animal, resourceProvider);
    }

    public void launchAnimalDetails() {
        dataStore.save(new AnimalDetailsUseCase(animal));
        eventBus.send(new ActivityEvent(this, AnimalDetailsActivity.class, false));
    }

    private void populateFields(Animal animal, ResourceProvider resourceProvider) {
        name.set(animal.getName());
        age.set(resourceProvider.getString(animal.getAge().getFormattedName()));

        setBreeds(animal.getBreedList());
        setPhoto(animal.getMedia());
        setDistance(animal.getContact());
    }

    private void setDistance(Contact contactInfo) {
        onLifecycle(locationHelper.getCurrentDistanceToContactInfo(contactInfo)
                                  .subscribe(this::displayDistance, Throwable::printStackTrace));
    }

    private void displayDistance(Integer returnDistance) {
        if (returnDistance == -1) {
            distanceVisibility.set(false);
        } else {
            distanceVisibility.set(true);
            String distanceString = returnDistance.toString();
            if (returnDistance == 0) {
                distanceString = "< 1";
            }
            distance.set(distanceString);
        }
    }

    private void setBreeds(List<String> breedList) {
        StringBuilder breedString = new StringBuilder();
        for (int i = 0; i < breedList.size(); i++) {
            String currentBreed = breedList.get(i);
            breedString.append(currentBreed);
            if (i != breedList.size() - 1) {
                breedString.append(" / ");
            }
        }
        this.breeds.set(breedString.toString());
    }

    private void setPhoto(Media media) {
        if (media != null && media.getPhotoList() != null && !media.getPhotoList().isEmpty()) {
            Photo finalPhoto = null;
            for (Photo photo : media.getPhotoList()) {
                if (photo.getSize() == PhotoSize.LARGE) {
                    finalPhoto = photo;
                    break;
                }
            }
            if (finalPhoto != null) {
                imageUrl.set(finalPhoto.getUrl());
            }
        }
    }

    public static class Factory {
        private ViewEventBus eventBus;
        private TransientDataStore dataStore;
        private ResourceProvider resourceProvider;
        private LocationHelper locationHelper;

        @Inject
        public Factory(ViewEventBus eventBus, TransientDataStore dataStore, ResourceProvider resourceProvider, LocationHelper locationHelper) {
            this.eventBus = eventBus;
            this.dataStore = dataStore;
            this.resourceProvider = resourceProvider;
            this.locationHelper = locationHelper;
        }

        public AnimalListItemViewModel newInstance(Animal animal) {
            return new AnimalListItemViewModel(animal, eventBus, dataStore, resourceProvider, locationHelper);
        }
    }
}

File: app/src/main/java/tech/pauly/findapet/discover/AnimalListItemViewModel.kt
package tech.pauly.findapet.discover

import android.databinding.ObservableBoolean
import android.databinding.ObservableField
import android.view.View
import tech.pauly.findapet.R
import tech.pauly.findapet.data.models.Animal
import tech.pauly.findapet.data.models.AnimalUrl
import tech.pauly.findapet.data.models.Sex
import tech.pauly.findapet.shared.BaseViewModel
import tech.pauly.findapet.shared.ResourceProvider
import tech.pauly.findapet.shared.SentencePlacement
import tech.pauly.findapet.shared.datastore.AnimalDetailsUseCase
import tech.pauly.findapet.shared.datastore.AnimalDetailsUseCase.Tab
import tech.pauly.findapet.shared.datastore.TransientDataStore
import tech.pauly.findapet.shared.events.ActivityEvent
import tech.pauly.findapet.shared.events.DialogEvent
import tech.pauly.findapet.shared.events.ViewEventBus
import tech.pauly.findapet.utils.ObservableString
import tech.pauly.findapet.utils.safeGet
import javax.inject.Inject

open class AnimalListItemViewModel(private val animal: Animal,
                                   private val eventBus: ViewEventBus,
                                   private val dataStore: TransientDataStore,
                                   private val resourceProvider: ResourceProvider) : BaseViewModel() {
    open var id: Int = 0
    var name = ObservableString()
    var imageUrl = ObservableField<AnimalUrl>()
    var age = ObservableString()
    var breeds = ObservableString()
    var distanceVisibility = ObservableBoolean(false)
    var warning = ObservableBoolean(false)

    private var sex = Sex.MISSING

    init {
        animal.also {
            id = it.id
            name.set(it.name)
            age.set(resourceProvider.getString(it.age.formattedName))
            breeds.set(it.formattedBreedList)
            imageUrl.set(animal.primaryPhotoUrl)
            warning.set(it.warning)
            sex = it.sex
        }
    }

    fun launchAnimalDetails(tab: Tab) {
        dataStore += AnimalDetailsUseCase(animal, tab)
        eventBus += ActivityEvent(this, AnimalDetailsActivity::class, false)
    }

    @Suppress("UNUSED_PARAMETER")
    fun showPetWarningDialog(v: View) {
        if (!warning.get()) {
            launchAnimalDetails(Tab.DETAILS)
            return
        }
        val bodyText = resourceProvider.getSexString(R.string.pet_warning_dialog_body,
                sex,
                name.safeGet(),
                SentencePlacement.OBJECT,
                SentencePlacement.SUBJECT)
        eventBus += DialogEvent(this,
                R.string.pet_warning_dialog_title,
                bodyText,
                R.string.pet_warning_dialog_contact,
                R.string.dialog_cancel,
                { launchAnimalDetails(Tab.CONTACT) },
                null,
                imageUrl.get(),
                R.color.warning)
    }

    open class Factory @Inject
    constructor(private val eventBus: ViewEventBus,
                private val dataStore: TransientDataStore,
                private val resourceProvider: ResourceProvider) {

        open fun newInstance(animal: Animal): AnimalListItemViewModel {
            return AnimalListItemViewModel(animal, eventBus, dataStore, resourceProvider)
        }
    }
}
